### DISCUSSION TITLE
Recommended way of handling rejected file permissions

### DISCUSSION THREAD

**Discussion by vinistock** [microsoft/vscode-discussions#1089] [Extension Development QnA] [answered]
### Context

Ruby dependencies (called gems) are typically installed in a directory outside of the workspace currently being worked on. Generally, they are installed somewhere in the user's home directory (e.g.: `~/.gem/ruby/3.3.0/gems`).

The [Ruby LSP language server](https://github.com/Shopify/ruby-lsp) needs to read the files under those directories, so that we can index the dependencies and keep a representation of all declarations available to that workspace.

However, if the user previously denied VS Code the permission to access the directories under the home directory, this fails because the Ruby LSP process will also not have permission to read the files.

We have the same problem when automatically discovering the Ruby version that a workspace is using. To do that, we depend on having access to version managers and the Ruby binary, which are both also often inside the user's home directory. We need to be able to run both executables so that we can discover the version.

### Questions

1. Is there a recommended way for handling this scenario where VS Code was not given permissions to a directory? Or is it mostly just checking for permissions using `vscode.workspace.fs.stat` and then showing error messages?
2. Is it possible to re-trigger the dialog asking for permissions from the user? Preferably with a custom message so that we can explain why we need them
3. If extensions can't show the permission dialog, then what's the recommended way to fix the permissions? Can this be done from inside VS Code? (So that we can provide instructions for users on what to do)

**Comment by abrilurena** [comment]
Hi, @vinistock! i hope you're doing great. It's been a while, do you still need help with this?

**Reply by vinistock** [reply]
Hey, @abrilurena! Hope you're doing great too, thank you.

Yes, I'd still appreciate some guidance.

**Comment by isidorn** [comment]
@bpasero can you please help our friends from Shopify. Thank you 🙏

**Comment by bpasero** [comment, marked as answer]
I am not aware of a solution here, sorry. There is https://github.com/microsoft/vscode/issues/48659 as feature request to deal with files that need elevation or are write protected, but I am not sure it would include this issue. 

Curious: if the user has denied access to the user home folder, I would assume a lot of other things fail to work too: for example, we store extensions in the home folder, so how does the user install extensions? Or are they using VS Code in portable mode?

**Reply by vinistock** [reply]
Thanks for the information. To be honest, I'm still not 100% sure myself of what steps were taken by the users to achieve this state.

We had some reports that our language server was only working properly on MacOS if you gave VS Code full disk access, which is certainly not necessary given that I didn't do that myself and everything works.

Based on that, I figured that the user must have denied some permission, but they weren't sure which exact permission. I'm still trying to understand what has to be denied to lead to that specific state and how we can check the permissions ahead of time to handle the error more gracefully.

Here's the original discussion if you're interested https://github.com/Shopify/ruby-lsp/issues/1586#issuecomment-1991931424.
